### ISSUE TITLE
Support custom logging handlers

### ISSUE THREAD

**Issue by djuelg** [quarkusio/quarkus#6133]
Related to #2409 

You closed the referenenced issue, since a Sentry extension has been provided.
Nevertheless, it would be good to have custom logging handlers like Wildfly does ([see here](https://docs.jboss.org/author/display/WFLY10/Handlers#Handlers-customhandler)) to support non-standard logging pipelines.

**Comment by hartimcwildfly** [comment]
adding a single custom handler (which extends the org.jboss.logmanager.ExtHandler class)
should be very easy as the already existing Handlers (console, syslog and file) also extend the org.jboss.logmanager.ExtHandler class. But if you'd like to link a logger to a specific handler, you have to add the ability to select a specific handler for a logger, which is more complex.

**Comment by hartimcwildfly** [comment]
Note:
In the WildFly world it's common to provide a custom handler which extends java.util.logging.Handler and which follows the JavaBean convention. (public noarg constructor, public setter methods). Then we can set the properties via quarkus config injected as Map<String, String>. For using the custom handler in native mode, it has to be annotated with @RegisterForReflection. I think this is a quite dirty solution. But if you want to configure your custom handler without reflection you have to write an extension for every handler you need or you configure the custom handler via a separate file or via environment properties.

tl;dr  If we want a configurable custom handler: 
 - do it via a separate file
 - do it via env vars
 - do it with generic properties in quarkus config and invoke the setters with reflection
 - (write an extension for every custom handler)

**Comment by dmlloyd** [comment]
The way I would solve this would be to add a "custom" handler config type with a nested map, as you say, but instead of using reflection, I would generate a configurator class that sets up the custom handler by directly invoking the setters on the handler object, using the configuration converters to convert as needed.  This avoids reflection problems but still allows the properties to be set at run time.  Care should be taken to ensure that unknown properties are properly logged.

**Comment by hartimcwildfly** [comment]
Sounds not that bad. Maybe I'm going to do this.

**Comment by hartimcwildfly** [comment]
Actually there is a possibility to use custom handlers. You just have to provide (/write an extension) a RuntimeValue<Optional<Handler>> like in io.quarkus.logging.gelf.GelfLogHandlerRecorder.
So there is only the problem with the linking of the handlers. I would propose to introduce a RuntimeValue<Entry<String, Handler>> for namedHandlers and an optional handlerName in CategoryConfig. So the named handlers would only be used if explicitly referenced in the CategoryConfig.

**Comment by hartimcwildfly** [comment]
Somebody already works on named handlers. See #5861.

**Comment by iustinov** [comment]
@dmlloyd, does the Quarkus team have plans for support aspectj (e.g. for logging)?

**Comment by hartimcwildfly** [comment]
aspectj-maven-plugin still does not support Java11 so I think they won't have planys for aspectj.

**Comment by dmlloyd** [comment]
No, there are currently no plans to support aspectj that I know of.

**Comment by iustinov** [comment]
It`s not very good. 
Does quarkus have any other way for aumatically log entry/exit of methods in application?

**Comment by hartimcwildfly** [comment]
You can use interceptors. See https://quarkus.io/guides/cdi-reference

**Comment by iustinov** [comment]
@hartimcwildfly, thank you! It works for me!

**Comment by hartimcwildfly** [comment]
I managed to write an extension which wraps my custom loghandler. But it requires some changes in quarkus core : a new BuildItem, as well as the LoggingSetupRecorder has to be adjusted. I did the changes from a branch forking the v1.2.1.Final. I will adapt them to the current master and open a PR.

**Comment by hartimcwildfly** [comment]
See PR #7568

**Comment by pkocandr** [comment]
@hartimcwildfly Hi! Could you please provide an example how to configure a custom log handler? I would like to test if it's possible to use Kafka log appender with this.

**Comment by hartimcwildfly** [comment]
Yes I can add some code where you just have to place your Appender. But please mind that the code will not be compileable.

**Comment by hartimcwildfly** [comment]
Or do you mean a custom **named** handler? (Refering to my PR #7568 ?)
Custom handlers are already supported.

**Comment by hartimcwildfly** [comment]
https://github.com/hartimcwildfly/quarkus-custom-loghandler

**Comment by hlavki** [comment]
Is this still open problem or PR #7568 is included in 1.5.0 release and this issue can be closed? thanks

**Comment by hlavki** [comment]
Would be great to update [Documentation](https://quarkus.io/guides/logging) about this feature.

**Comment by hartimcwildfly** [comment]
Sorry, I can't do it in the near future. But as this is not a functionality which can be used directly by the user, I don't think it's too bad that there is no documentation. Because for using a named custom logging handler the user has to write his own quarkus extension which provides a NamedLogHandlersBuildItem. (same as when the wants to use an unnamed custom loghandler where one needs to provide a LogHandlerBuildItem.) So one has to look at the quarkus code anyway.

**Comment by GustavoGramajo** [comment]
@hartimcwildfly i wonder if i could use what you did to grab all logs from my app and send it to a queue (ActiveMQ). As far as i can tell, it doesn't seem to be a simpler way to do something like this.

**Comment by hartimcwildfly** [comment]
If you want to grab ALL your logs you have to provide a LogHandlerBuildItem (which was possible before my PR). If you want to use named handlers you have to provide a NamedLogHandlersBuildItem which can contain none, one or more handlers.

**Comment by GustavoGramajo** [comment]
@hartimcwildfly Does it mean that i should write my own extension? Will i still be able to also print to the console?

**Comment by hartimcwildfly** [comment]
Yes you have to write your own extension. And you are still able to use other log handlers.

**Comment by debu999** [comment]
> aspectj

https://github.com/dev-aspectj/aspectj-maven-plugin is supporting java 17. Is there any plans on this with quarkus. 
We have to write multiple loggers for cdi and non cdi java classes and methods. It becomes difficult for code generators viz. interfaces in rest easy reactive etc.

**Comment by debu999** [comment]
in metrics we have count/max/avg but not the total
